server/server-api/src/main/java/software/amazon/smithy/java/server/FilteredService.java

```diff
@@ -0,0 +1,93 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import software.amazon.smithy.java.core.schema.Schema;
+import software.amazon.smithy.java.core.schema.SerializableStruct;
+import software.amazon.smithy.java.core.serde.TypeRegistry;
+import software.amazon.smithy.java.framework.model.UnknownOperationException;
+
+/**
+ * A service implementation that filters operations based on a provided filter.
+ * <p>
+ * This class wraps an existing service and applies an operation filter to control
+ * which operations are exposed through this service.
+ */
+public final class FilteredService implements Service {
+
+    private final Service delegate;
+    private final OperationFilter filter;
+    private final Map<String, Operation<? extends SerializableStruct, ? extends SerializableStruct>> filteredOperations;
+    private final List<Operation<? extends SerializableStruct, ? extends SerializableStruct>> operationList;
+
+    /**
+     * Creates a new filtered service by applying the specified filter to the delegate service.
+     *
+     * @param delegate The underlying service to filter
+     * @param filter The filter to apply to the service's operations
+     */
+    public FilteredService(Service delegate, OperationFilter filter) {
+        this.delegate = delegate;
+        this.filter = filter;
+        this.filteredOperations = delegate.getAllOperations()
+                .stream()
+                .filter(filter::include)
+                .collect(Collectors.toMap(Operation::name, Function.identity()));
+        this.operationList = new ArrayList<>(filteredOperations.values());
+    }
+
+    /**
+     * Retrieves an operation by name, if it exists and passes the filter.
+     *
+     * @param operationName The name of the operation to retrieve
+     * @return The filtered operation with the specified name
+     * @throws UnknownOperationException if no matching operation is found or it's filtered out
+     */
+    @Override
+    public <I extends SerializableStruct,
+            O extends SerializableStruct> Operation<I, O> getOperation(String operationName) {
+        var operation = filteredOperations.get(operationName);
+        if (operation == null) {
+            throw UnknownOperationException.builder().message("No matching operations found for request").build();
+        }
+        return (Operation<I, O>) operation;
+    }
+
+    /**
+     * Returns all operations that pass the filter.
+     *
+     * @return A list of all operations included by the filter
+     */
+    @Override
+    public List<Operation<? extends SerializableStruct, ? extends SerializableStruct>> getAllOperations() {
+        return operationList;
+    }
+
+    /**
+     * Returns the schema from the delegate service.
+     *
+     * @return The schema of the delegate service
+     */
+    @Override
+    public Schema schema() {
+        return delegate.schema();
```

Comment:
This schema will contain operations that are not returned by `getAllOperations()`. It's not a problem for now but I can see this violating some kind of assumption down the line.